SMLHelper/Options/Attributes/ConfigFileMetadata.cs

```diff
@@ -0,0 +1,628 @@
+#elif BELOWZERO
+    using Text = TMPro.TextMeshProUGUI;
+#endif
+#if SUBNAUTICA_STABLE
+    using Oculus.Newtonsoft.Json;
+#else
+    using Newtonsoft.Json;
+#endif
+
+    internal class ConfigFileMetadata<T> where T : ConfigFile, new()
+    {
+        public T Config { get; } = new T();
+
+        public IQMod QMod { get; } = QModServices.Main.GetMod(Assembly.GetAssembly(typeof(T)));
+
+        /// <summary>
+        /// The <see cref="MenuAttribute"/> relating to this <see cref="ModOptions"/> menu.
+        /// </summary>
+        public MenuAttribute MenuAttribute { get; private set; }
+
+        /// <summary>
+        /// A dictionary of <see cref="ModOptionAttributeMetadata{T}"/>, indexed by <see cref="ModOption.Id"/>.
+        /// </summary>
+        public Dictionary<string, ModOptionAttributeMetadata<T>> ModOptionAttributesMetadata { get; private set; }
+
+        /// <summary>
+        /// Process metadata for members of <typeparamref name="T"/>.
+        /// </summary>
+        public void ProcessMetadata()
+        {
+            Stopwatch stopwatch = new Stopwatch();
+
+            if (Logger.EnableDebugging)
+                stopwatch.Start();
+
+            MenuAttribute = typeof(T).GetCustomAttribute<MenuAttribute>(true) ?? new MenuAttribute(QMod.DisplayName);
+            ModOptionAttributesMetadata = new Dictionary<string, ModOptionAttributeMetadata<T>>();
+
+            processMetadata();
+
+            if (Logger.EnableDebugging)
+            {
+                stopwatch.Stop();
+                Logger.Debug($"[{QMod.DisplayName}] [{typeof(T).Name}] OptionsMenuBuilder metadata parsed in {stopwatch.ElapsedMilliseconds}ms.");
+            }
+        }
+
+        private void processMetadata()
+        {
+            var bindingFlags = BindingFlags.Instance | BindingFlags.Public;
+            foreach (PropertyInfo property in typeof(T).GetProperties(bindingFlags)
+                .Where(memberIsDeclaredInConfigFileSubclass) // Only care about members declared in a subclass of ConfigFile
+                .Where(memberIsNotIgnored)) // Filter out explicitly ignored members
+            {
+                processFieldOrProperty(property, MemberType.Property, property.PropertyType);
+            }
+
+            foreach (FieldInfo field in typeof(T).GetFields(bindingFlags)
+                .Where(memberIsDeclaredInConfigFileSubclass) // Only care about members declared in a subclass of ConfigFile
+                .Where(memberIsNotIgnored)) // Filter out explicitly ignored members
+            {
+                processFieldOrProperty(field, MemberType.Field, field.FieldType);
+            }
+
+            foreach (MethodInfo method in typeof(T).GetMethods(bindingFlags | BindingFlags.Static)
+                .Where(memberIsDeclaredInConfigFileSubclass) // Only care about members declared in a subclass of ConfigFile
+                .Where(memberIsNotIgnored)) // Filter out explicitly ignored members
```

Comment:
`BindingFlags.Static`  
Will this only pick up static methods?

Reply:
With BindingFlags, only flags that are provided will be looked at.

```cs
var bindingFlags = BindingFlags.Instance | BindingFlags.Public;
```
```cs
foreach (MethodInfo method in typeof(T).GetMethods(bindingFlags | BindingFlags.Static)
```
Resulting in a combination of `Instance`, `Public` and `Static`, so only public methods will be looked at, and they can be either Instance or Static.